server/controllers/roomController.ts

```diff
@@ -0,0 +1,122 @@
+import { Room, RoomClass } from "../db/room";
+import { Player, PlayerClass } from "../db/player";
+import { DocumentType } from "@typegoose/typegoose";
+import { Card } from "../db/card";
+
+function shuffle<T>(array: T[]): T[] {
+  var currentIndex = array.length,
+    temporaryValue,
+    randomIndex;
+
+  // While there remain elements to shuffle...
+  while (0 !== currentIndex) {
+    // Pick a remaining element...
+    randomIndex = Math.floor(Math.random() * currentIndex);
+    currentIndex -= 1;
+
+    // And swap it with the current element.
+    temporaryValue = array[currentIndex];
+    array[currentIndex] = array[randomIndex];
+    array[randomIndex] = temporaryValue;
+  }
+
+  return array;
+}
+
+export class RoomController {
+  static HAND_SIZE = 12;
+  /**
+   * Removes a {@link Player} from any {@link Room}s they are part of, given only their connectionId
+   * @param connectionId The connectionId of the {@link Player} to remove.
+   */
+  async removePlayer(
+    connectionId: string
+  ): Promise<DocumentType<RoomClass>[] | undefined> {
+    const player = await Player.findOne({ connectionId });
+    if (!player) {
+      console.error(
+        `No player with connection ID ${connectionId} was found to remove from rooms.`
+      );
+      return;
+    }
+    // Get IDs of the rooms affected
+    const affectedRooms = await Room.find({ players: { $all: [player._id] } });
+    const affectedRoomIds = affectedRooms.map((room) => room._id);
+
+    await Room.updateMany(
+      { players: { $all: [player._id] } },
+      { $pull: { players: player._id } }
+    );
+
+    await Player.deleteMany({ connectionId }).exec();
+
+    return await Room.find({ _id: { $in: affectedRoomIds } }).populate(
+      "players"
+    );
+  }
+  /**
+   * Creates and returns a new {@link Room}.
+   * @param open Whether or not the {@link Room} should be accessible by the public
+   */
+  public async createRoom(open: boolean = false) {
+    // Just in case there are collisions when generating the code, retry until no collisions.
+    while (true) {
+      try {
+        return await Room.create({ open });
+      } catch (err) {
+        console.error(err);
+      }
+    }
+  }
+
+  /**
+   * Joins a {@link Room} if it exists. Rejects if the {@link Room} does not exist.
+   * @param roomCode The code of the {@link Room} to join
+   * @param connectionId The ID of the connection the player is joining on
+   * @param playerName The name of the player within the {@link Room}
+   */
+  public async joinRoom(
+    roomCode: string,
+    connectionId: string,
+    playerName: string
+  ): Promise<{
+    player: DocumentType<PlayerClass>;
+    room: DocumentType<RoomClass>;
+  }> {
+    const player = await Player.create({
+      connectionId: connectionId,
+      name: playerName,
```

Comment:
Are there any restrictions to the player name? Any Unicode, any length? Should we limit length in case someone dumps a novel?

Reply:
I guess we could, but then we couldn't have the entirety of the Great Gatsby